File: 0067-add-binary/67.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        len_a, len_b = len(a), len(b)
        max_len = max(len_a, len_b)
        res = ['0'] * (max_len + 1)  # +1 for potential carry
        i, j, k = len_a - 1, len_b - 1, max_len
        carry = 0

        while i >= 0 or j >= 0 or carry:
            total = carry
            if i >= 0:
                total += int(a[i])
                i -= 1
            if j >= 0:
                total += int(b[j])
                j -= 1
            res[k] = str(total % 2)
            carry = total // 2
            k -= 1

        return ''.join(res[k+1:])  # Skip leading zero if unused

File: 0067-add-binary/67.rs
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let (len_a, len_b) = (a.len(), b.len());
        let max_len = len_a.max(len_b);
        let mut res = vec!['0'; max_len + 1];
        let (a_bytes, b_bytes) = (a.as_bytes(), b.as_bytes());
        let (mut i, mut j, mut k) = (len_a as i32 - 1, len_b as i32 - 1, max_len as i32);
        let mut carry = 0;

        while i >= 0 || j >= 0 || carry > 0 {
            let mut total = carry;
            if i >= 0 {
                total += (a_bytes[i as usize] - b'0') as u8;
                i -= 1;
            }
            if j >= 0 {
                total += (b_bytes[j as usize] - b'0') as u8;
                j -= 1;
            }
            res[k as usize] = ((total % 2) + b'0') as char;
            carry = total / 2;
            k -= 1;
        }

        // Skip leading zero if not needed
        res[(k + 1) as usize..].iter().collect()
    }
}

